kernel/src/net/socket/util/message_header.rs

```diff
@@ -1,32 +1,190 @@
+            reader.skip(padding_len);
+        }
+
+        Ok(msgs)
+    }
+
+    fn read_from(header: &CControlHeader, reader: &mut VmReader) -> Result<Option<Self>> {
+        let Some(level) = header.level() else {
+            warn!("unsupported control message level in {:?}", header);
+            reader.skip(header.payload_len());
+            return Ok(None);
+        };
+
+        match level {
+            CSocketOptionLevel::SOL_SOCKET => {
+                // Linux manual pages say (https://man7.org/linux/man-pages/man7/unix.7.html):
+                // "For historical reasons, the ancillary message types listed below are specified
+                // with a SOL_SOCKET type even though they are AF_UNIX specific."
+                let msg = UnixControlMessage::read_from(header, reader)?;
+                Ok(msg.map(Self::Unix))
+            }
+            _ => {
+                warn!("unsupported control message level in {:?}", header);
+                reader.skip(header.payload_len());
+                Ok(None)
+            }
+        }
+    }
+
+    pub fn write_all_to(msgs: &[Self], writer: &mut VmWriter) -> usize {
+        let mut len = 0;
+
+        for msg in msgs.iter() {
+            let header = match msg.write_to(writer) {
+                Ok(header) => header,
+                // This occurs when the buffer is too short or when some page faults cannot be
+                // handled. However, at this point, there is no good way to report the errors to
+                // user space. According to the Linux implementation, it seems okay to silently
+                // ignore errors here.
+                Err(_) => {
+                    warn!("setting MSG_CTRUNC is not supported");
+                    break;
+                }
+            };
+
+            len += header.total_len();
+
+            let padding_len = header.padding_len().min(writer.avail());
+            writer.skip(padding_len);
+            len += padding_len;
+        }
+
+        len
+    }
+
+    fn write_to(&self, writer: &mut VmWriter) -> Result<CControlHeader> {
+        match self {
+            Self::Unix(msg) => msg.write_to(writer),
+        }
+    }
+}
+
+/// `cmsghdr` in Linux.
+///
+/// Reference: <https://elixir.bootlin.com/linux/v6.13/source/include/uapi/linux/netlink.h#L52>.
+#[repr(C)]
+#[derive(Debug, Clone, Copy, Pod)]
+pub struct CControlHeader {
+    /// Data byte count, including hdr
+    len: usize,
+    /// Originating protocol
+    level: i32,
+    /// Protocol-specific type
+    type_: i32,
+}
+
+/// Alignment of control messages.
+///
+/// Reference: <https://elixir.bootlin.com/linux/v6.13/source/include/linux/socket.h#L119>.
+const CMSG_ALIGN: usize = size_of::<usize>();
+
+impl CControlHeader {
+    /// Creates a control message header with the level, type, and payload length.
+    pub fn from_payload_len(level: CSocketOptionLevel, type_: i32, payload_len: usize) -> Self {
```

Comment:
Naming this method `new` is appropriate. This method accepts additional parameters beyond the payload length, and the length specifically refers to the payload length, which has emphasized by the parameter name.